### PR TITLE
feat: elevation profile

### PR DESCRIPTION
**Pull request by totorototo**
whatsapp image 2018-02-20 at 21 20 19

Do not merge yet!

### CODE REVIEW

File: app/components/common/cardList/CardList.js [hunk 1]
```diff
@@ -34,23 +34,23 @@ export default class CardList extends Component {
   render() {
     return (
       <Card title={this.props.title}>
-        {this.props.list.map(item =>
+        {this.props.list.map(item => (
```
Comment (guipasmoi) [inline app/components/common/cardList/CardList.js:37]: PRETTIER CHANGED?

File: app/components/specific/cards/AthleteCard.js [hunk 2]
```diff
@@ -9,14 +9,20 @@ export default class AthleteCard extends Component {
...
 
   render() {
     const { athlete } = this.props;
     return (
-      <Card title="ATHLETE" image={{ uri: athlete.profile }}>
+      <Card
+        title="ATHLETE"
+        image={{
+          uri:
+            "https://blog.strava.com/wp-content/uploads/2017/12/Strava_YearInStats_Header_0_1x-1.jpg"
```
Comment (guipasmoi) [inline app/components/specific/cards/AthleteCard.js:23]: should'nt be here

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  static createAreaGraph(edges, graphWidth, graphHeight) {
+    // Create our x-scale.
+    const scaleX = ElevationProfile.createXScale(
+      0,
+      positionHelper.computeDistance(...edges),
+      graphWidth
+    );
+
+    // Collect all y values.
+    const altitudes = edges.reduce(
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js]: this is a map

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+const { Surface, Shape } = ART;
+const d3 = {
+  scale,
+  shape,
+  d3Array
+};
+
+export default class ElevationProfile extends Component {
+  static propTypes = {
+    path: PropTypes.shape({
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js:52]: somthing is too much here?

File: app/components/specific/map/RaceMap.js [hunk 5]
```diff
@@ -61,16 +70,24 @@ export default class RaceMap extends Component {
     const { race, clubMembers } = this.props;
 
     const sampleRegion = {
-      latitude: race.path.coordinates[0].latitude,
-      longitude: race.path.coordinates[0].longitude,
+      latitude: race.path.edges[0].src.latitude,
```
Comment (guipasmoi) [inline app/components/specific/map/RaceMap.js]: what? why 0?

File: app/components/specific/map/RaceMap.js [hunk 6]
```diff
@@ -61,16 +70,24 @@ export default class RaceMap extends Component {
...
+      longitude: race.path.edges[0].src.longitude,
       latitudeDelta: LATITUDE_DELTA,
       longitudeDelta: LONGITUDE_DELTA
     };
 
+    const raceData = race.path.edges.reduce(
+      (locations, edge) => [
```
Comment (guipasmoi) [inline app/components/specific/map/RaceMap.js:82]: last point is missing

File: app/store/sagas/race.js [hunk 7]
```diff
@@ -4,16 +4,32 @@ import { SET_CURRENT_USER_ID } from "../constants/actionTypes";
 import { setCurrentRaceID } from "../actions/data";
 import { setEntity } from "../actions/entities";
 import { markers, coordinates } from "./data/data";
+import positionHelper from "../services/helpers/gps";
 
 function* listRaces() {
-  // 1- mock strava api call
+  const edges = coordinates.reduce((accu, location, index) => {
```
Comment (guipasmoi) [inline app/store/sagas/race.js]: this shouldn't be in saga

Comment (guipasmoi) [reply app/store/sagas/race.js]: extract service

File: app/components/specific/map/RaceMap.js (hunk 6 above)
Comment (totorototo) [reply app/components/specific/map/RaceMap.js:82]: It was just to test how good reviewer you were! 👍

File: app/store/sagas/race.js (hunk 7 above)
Comment (totorototo) [reply app/store/sagas/race.js]: roger that! It makes sense!

File: app/components/common/cardList/CardList.js (hunk 1 above)
Comment (totorototo) [reply app/components/common/cardList/CardList.js:37]: I guess.

Comment (guipasmoi) [reply app/components/common/cardList/CardList.js:37]: try running prettier on every file to check it

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 8]
```diff
@@ -53,16 +71,81 @@ export default class ElevationProfile extends Component {
...
+      return ELEVATION_GRADE.DIFFICULT;
+    } else if (Math.abs(percent) >= 15) {
+      return ELEVATION_GRADE.HARD;
+    }
+    return ELEVATION_GRADE.UNKNOWN;
+  };
+
+  // TODO: to be moved
+  static convertGradeToColor = grade => {
+    if (grade === ELEVATION_GRADE.SMALL) {
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js]: return ELEVATION_COLORS[grade]

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 9]
```diff
@@ -53,16 +71,81 @@ export default class ElevationProfile extends Component {
...
+        distanceDone,
+        grade,
+        color
+      };
+      return [...accu, enhancedEdge];
+    }, []);
+  };
+
+  // TODO: to be moved
+  static groupBy = (...edges) => {
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js]: groupBy should be prototypeOn array and take a grouping fct that return the key of an item

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 10]
```diff
@@ -53,16 +71,81 @@ export default class ElevationProfile extends Component {
...
+      currentGrade = edge.grade;
+      return [...accu, [edge]];
+    }, []);
+  };
 
   static createAreaGraph(edges, graphWidth, graphHeight) {
-    // TODO: create area foreach partition
-    // const partitions = positionHelper.partitionPath(...edges);
+    const updatedEdges = ElevationProfile.addData(...edges);
+    const sortedEdges = ElevationProfile.groupBy(...updatedEdges);
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js]: misleading name

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 11]
```diff
@@ -53,16 +71,81 @@ export default class ElevationProfile extends Component {
...
+      }
+      currentGrade = edge.grade;
+      return [...accu, [edge]];
+    }, []);
+  };
 
   static createAreaGraph(edges, graphWidth, graphHeight) {
-    // TODO: create area foreach partition
-    // const partitions = positionHelper.partitionPath(...edges);
+    const updatedEdges = ElevationProfile.addData(...edges);
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js]: i kind of dont like that addData

File: app/store/services/event.js [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+      index,
+      distanceDone,
+      grade,
+      color
+    };
+    return [...accu, enhancedEdge];
+  }, []);
+};
+
+const groupBy = (...edges) => {
```
Comment (guipasmoi) [inline app/store/services/event.js]: duplicate?

File: app/store/services/helpers/event.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
+import { ELEVATION_COLORS, ELEVATION_GRADE } from "../../constants/elevation";
+
+export const convertPercentToGrade = percent => {
+  if (Math.abs(percent) < 5) {
```
Comment (guipasmoi) [inline app/store/services/helpers/event.js]: write fct isInRange

File: app/components/specific/elevationProfile/ElevationProfile.js (hunk 10 above)
Comment (totorototo) [reply app/components/specific/elevationProfile/ElevationProfile.js]: Cleaning in progress!

File: app/components/specific/elevationProfile/ElevationProfile.js (hunk 3 above)
Comment (totorototo) [reply app/components/specific/elevationProfile/ElevationProfile.js]: Could be a reduce as well!

File: app/store/sagas/race.js (hunk 7 above)
Comment (totorototo) [reply app/store/sagas/race.js]: Done

File: app/store/services/event.js (hunk 12 above)
Comment (totorototo) [reply app/store/services/event.js]: Cleaning in progress! At the end, only one will....

File: app/components/specific/elevationProfile/ElevationProfile.js (hunk 9 above)
Comment (totorototo) [reply app/components/specific/elevationProfile/ElevationProfile.js]: 👍

File: app/components/specific/elevationProfile/ElevationProfile.js (hunk 11 above)
Comment (totorototo) [reply app/components/specific/elevationProfile/ElevationProfile.js]: Me neither. This should be removed, rename asap

File: app/store/services/helpers/event.js (hunk 13 above)
Comment (totorototo) [reply app/store/services/helpers/event.js]: 👍👍👍

File: app/components/specific/elevationProfile/ElevationProfile.js (hunk 11 above)
Comment (totorototo) [reply app/components/specific/elevationProfile/ElevationProfile.js]: clean up done

File: app/components/common/cardList/CardList.js (hunk 1 above)
Comment (totorototo) [reply app/components/common/cardList/CardList.js:37]: prettier indeed

File: app/components/specific/elevationProfile/ElevationProfile.js (hunk 8 above)
Comment (totorototo) [reply app/components/specific/elevationProfile/ElevationProfile.js]: obselete - deprecated! using d3 to deal with colors, range and threshold.

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 14]
```diff
@@ -15,22 +16,42 @@ const d3 = {
...
 };
 
-const ELEVATION_GRADE = {
-  SMALL: 0,
-  EASY: 1,
-  MEDIUM: 2,
-  DIFFICULT: 3,
-  HARD: 4,
-  UNKNOWN: 5
+const getRange = percent => {
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js]: i think u missunderstood me
const isInRange => (lowerValue, uperValue) => currentValue =>{
  return lowerValue<=currentValue && currentValue < uperValue
}

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 15]
```diff
@@ -15,22 +16,42 @@ const d3 = {
...
+  const val = getRange(item.percent);
+  // eslint-disable-next-line no-param-reassign
+  accu[val] = accu[val] || [];
+  accu[val].push(item);
+  return accu;
+};
+
+const groupEdgesBySequence = function(accu, value, index, collection) {
+  if (index > 0 && value.index - collection[index - 1].index === 1) {
+    const group = last(accu);
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js]: currentGroup ?

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 16]
```diff
@@ -15,22 +16,42 @@ const d3 = {
...
-const ELEVATION_COLORS = {
-  SMALL: "#f4f6f5",
-  EASY: "#ECBC3E",
-  MEDIUM: "#EA8827",
-  DIFFICULT: "#E1351D",
-  HARD: "#96451F",
-  UNKNOWN: "#00451F"
+// eslint-disable-next-line no-extend-native
+Array.prototype.groupBy = function(fn, accu) {
+  return this.reduce(fn, accu);
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js]: WTF !! that s reduce not groupBy

Comment (totorototo) [reply app/components/specific/elevationProfile/ElevationProfile.js]: I guess you do not get my point! 
groupEdgesByRange - groupEdgesBySequence functions are doing that job! 
And those functions are groupeBy first parameter.....

File: app/components/specific/elevationProfile/ElevationProfile.js (hunk 15 above)
Comment (totorototo) [reply app/components/specific/elevationProfile/ElevationProfile.js]: yep! but tu pinailles!

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 17]
```diff
@@ -31,27 +32,15 @@ const getRange = percent => {
...
-    accu.push([value]);
-  }
-  return accu;
-};
-
 // eslint-disable-next-line no-extend-native
-Array.prototype.groupBy = function(fn, accu) {
-  return this.reduce(fn, accu);
+Array.prototype.groupBy = function(fn) {
+  return this.reduce((accu, item, index, array) => {
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js]: VERY GOOD 🥇  😃

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 18]
```diff
@@ -100,11 +89,20 @@ export default class ElevationProfile extends Component {
...
-
-    const groupedEdgesBySequence = {};
-    Object.entries(groupedEdgesByRange).forEach(([range, section]) => {
-      groupedEdgesBySequence[range] = section.groupBy(groupEdgesBySequence, []);
+    const groupedEdgesByRange = edges.groupBy(item => getRange(item.percent));
+
+    // eslint-disable-next-line
+    const data = Object.entries(groupedEdgesByRange).map(([grade, section]) => {
+      const indices = section.map(item => item.index);
+      const missingValueIndices = xor(
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js:141]: those lines have the smell of dark magic

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 19]
```diff
@@ -121,31 +119,23 @@ export default class ElevationProfile extends Component {
...
-          color: colorScale(grade)
-        };
-      })
-    );
-
-    const paths = Object.values(groupedPaths).map(sequences =>
-      sequences.map(item => item)
-    );
-    return flatten(paths);
+    return Object.entries(data).map(([grade, section]) => {
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js]: 😄

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 20]
```diff
@@ -162,7 +152,7 @@ export default class ElevationProfile extends Component {
               d={area.path}
               stroke={area.color}
               fill={area.color}
-              strokeWidth={1}
+              strokeWidth={0.2}
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js]: lol issue is less visible like that !! no?

File: app/components/specific/elevationProfile/ElevationProfile.js (hunk 17 above)
Comment (totorototo) [reply app/components/specific/elevationProfile/ElevationProfile.js]: I miss my reduce! 😋

File: app/components/specific/elevationProfile/ElevationProfile.js (hunk 18 above)
Comment (totorototo) [reply app/components/specific/elevationProfile/ElevationProfile.js:141]: No dark magic! 
In order to display area correctly we need to add data.

File: app/components/specific/elevationProfile/ElevationProfile.js (hunk 20 above)
Comment (totorototo) [reply app/components/specific/elevationProfile/ElevationProfile.js]: Smart, isn't it! 😂🤣

**Comment by totorototo** [comment]
some parts have to be reviewed asap

File: app/components/specific/elevationProfile/ElevationProfile.js [hunk 21]
```diff
@@ -85,7 +88,13 @@ export default class ElevationProfile extends Component {
     return d3.scale
       .scaleThreshold()
       .domain([1, 2, 3, 4])
-      .range(["#F4F6F5", "#ECBC3E", "#EA8827", "#E1351D", "#96451F"]);
+      .range([
```
Comment (guipasmoi) [inline app/components/specific/elevationProfile/ElevationProfile.js]: .range(Objec.values(ELEVATION_COLORS))

File: app/components/specific/cards/ClubActivitiesCard.js [hunk 22]
```diff
@@ -19,7 +21,7 @@ export default class ClubActivitiesCard extends Component {
         title="ACTIVITIES"
         list={activities.map((activity, index) => ({
           key: index,
-          image: activity.athlete.profile,
+          image: { name: getIconName("runner"), color: theme.PrimaryColor },
```
Comment (guipasmoi) [inline app/components/specific/cards/ClubActivitiesCard.js:24]: -_- that line is suspicious

File: app/routes/main/clubFeed/helper.js [hunk 23]
```diff
@@ -7,6 +7,7 @@ const mapping = {
   elevation: "landscape",
   duration: "timer",
   pace: "flash-on",
+  runner: "directions-run",
```
Comment (guipasmoi) [inline app/routes/main/clubFeed/helper.js:10]: what ???

File: app/store/sagas/clubs.js [hunk 24]
```diff
@@ -1,18 +1,31 @@
...
   listClubMembers,
   listClubActivities,
   filterClubAthletes
 } from "../services/clubs";
 import { getRankings } from "../services/activities";
 
-function* listActivities(clubID, membersIDs) {
+function* listActivities() {
+  const clubID = yield select(getCurrentClubID);
+  if (clubID === "loading") return;
```
Comment (guipasmoi) [inline app/store/sagas/clubs.js:24]: it s strange clubID === "loading"

File: app/store/services/clubs.js [hunk 25]
```diff
@@ -4,6 +4,9 @@ import { pick } from "lodash";
 import { API_ENDPOINT, RESOURCES, METHODS } from "../constants/rest";
 import { callJSONApi } from "./helpers/api";
 
+const computeID = (min, max) =>
+  Math.floor(Math.random() * (max - min + 1)) + min;
```
Comment (guipasmoi) [inline app/store/services/clubs.js:8]: wtf is that +1

File: .gitignore [hunk 26]
```diff
@@ -45,6 +45,7 @@ buck-out/
 # For more information about the recommended setup visit:
 # https://docs.fastlane.tools/best-practices/source-control/
 
+<<<<<<< ours
```
Comment (guipasmoi) [inline .gitignore]: fail

File: .gitignore [hunk 27]
```diff
@@ -56,11 +54,7 @@ fastlane/screenshots
...
 
-=======
-*/fastlane/report.xml
-*/fastlane/Preview.html
-*/fastlane/screenshots
 
 # Bundle artifact
 *.jsbundle
->>>>>>> theirs
+
```
Comment (guipasmoi) [inline .gitignore]: fix fail 😄

File: app/routes/main/clubFeed/helper.js (hunk 23 above)
Comment (totorototo) [reply app/routes/main/clubFeed/helper.js:10]: icon resource name I needed

File: app/store/sagas/clubs.js (hunk 24 above)
Comment (totorototo) [reply app/store/sagas/clubs.js:24]: We did (you did) the same for all application loading state!

File: app/store/services/clubs.js (hunk 25 above)
Comment (totorototo) [reply app/store/services/clubs.js:8]: random shit function I needed due to strava breaking api change! 
+1 is my legacy 🌮

File: app/components/specific/cards/ClubActivitiesCard.js (hunk 22 above)
Comment (totorototo) [reply app/components/specific/cards/ClubActivitiesCard.js:24]: again, strava breaking change api! 
We have no longer access to club athlete details...

**Pull request merged**